### DISCUSSION TITLE
[outdated] My experience trying to `bun build`

### DISCUSSION THREAD

**Discussion by llimllib** [oven-sh/bun#2936] [General]
I liked the idea of distributing a binary of [our CLI tool](https://github.com/readmeio/rdme), so I thought I'd give `bun build` a shot.

I'm giving my impressions as an attempted new user in the hope that they're helpful.

- I was surprised that bun installed to `~/.bun` instead of `/usr/local/bin`or `~/.local/bin` (where I relocated it to) #1678 
- I created a `build.js` as per [the documentation](https://bun.sh/docs/cli/build), but the docs did not tell me how to actually _execute_ the file
  - the answer is that you run `bun build.js`, but it took me lots of tries to figure that out
  - largely, it took me lots of tries because `bun build.js` does not throw any errors, even though the equivalent `bun build` command does
  - `bun build.js` not only fails with no errors, it returns a success error code
- `bun build src/cli.ts --outdir build` throws lots of errors about node.js modules:
```
error: Could not resolve Node.js builtin: "child_process".

const {spawn} = require('child_process');
                        ^
/Users/llimllib/readme/rdme/bun-test/node_modules/update-notifier-cjs/index.js:2:25 38
```

In a hello world project, I confirmed that that `require` works correctly.

At that point, I gave up; I'll wait until bun is more mature.

Just want to give my experience in the hope that it will be helpful. Good luck, I look forward to the future of bun.

**Comment by Jarred-Sumner** [comment]
Thank you for the feedback

> bun build src/cli.ts --outdir build throws lots of errors about node.js modules:

Can you pass `--target=node` or `--target=bun` and try again? We should change this error message to mention this flag should be added when a Node.js builtin module is imported or required. By default, `bun build` builds for browsers since in Bun's runtime we don't want to encourage asking people to run a build step.

> bun build.js not only fails with no errors, it returns a success error code

It should have a message in `logs` containing a `ResolveMessage` with the error you saw. Definitely a bug if `success` was `true` though.

**Reply by llimllib** [reply]
Thanks for the response! sorry I didn't get it until now.

> Can you pass --target=node

Thanks, that gets compilation down to one error, it doesn't like our `.d.ts` file that declares a module for an npm package that doesn't have types. Compiling via the `build.js` file still gives no errors and returns a `0` success code. 

I don't see a `logs` directory anywhere, is there somewhere I should look for that?

```
$ cat src/.sink.d.ts
// These packages don't have any TS types so we need to declare a module in order to use them.
declare module 'editor';
declare module 'update-notifier-cjs';

$ bun build src/cli.ts --outdir build --target=node
error: No matching export in "src/.sink.d.ts" for import "default"

import editor from 'editor';
       ^
/Users/llimllib/readme/rdme/bun-test/src/cmds/docs/edit.ts:9:8 193

$ cat build.js
await Bun.build({
  entrypoints: ['./src/cli.ts'],
  outdir: './build',
  target: 'node',
});

$ bun build.js

$ echo $?
0
```

**Comment by llimllib** [comment]
Oh now I see it, I never got that far down the documentation page!

- It seems like a very poor UI to not print errors by default! That will trip up a lot of people, not just me
- Shouldn't `bun` exit with a non-zero exit code if the build was a failure?

```
$ cat build.js
const result = await Bun.build({
  entrypoints: ['./src/cli.ts'],
  outdir: './build',
  target: 'node',
});

if (!result.success) {
  console.error('Build failed');
  for (const message of result.logs) {
    // Bun will pretty print the message object
    console.error(message);
  }
}

$ bun build.js
Build failed


error: No matching export in "src/.sink.d.ts" for import "default"

import editor from 'editor';
       ^

$ echo $?
0
```

**Reply by llimllib** [reply]
Removing `.sink.d.ts` gets me a successful compile:

```
$ bun build --compile src/index.ts --outfile bun-index --target=node
  [44ms]  bundle  975 modules
 [103ms] compile  bun-index
```

but it segfaults:

```
$ ./bun-index
Segmentation fault: 11
```

(I assumed I'd be able to add `compile: true` to my `Bun.build` command, but I couldn't find any appropriate option there so I figured it may not be implemented yet)

**Reply by llimllib** [reply]
that ^^^ was with bun 0.6.1, but I upgraded to 0.6.3 to see if it was something that had been fixed and no luck

**Reply by llimllib** [reply]
[here's the crash report from console.app](https://gist.github.com/llimllib/cce67a3e8b7c0cca3c47bd8d21d2e98c)

**Reply by llimllib** [reply]
I tried to attach `lldb`, but no joy:

```
$ lldb ./bun-index
(lldb) target create "./bun-index"
Current executable set to '/Users/llimllib/readme/rdme/bun-test/bun-index' (arm64).
(lldb) run
error: process exited with status -1 (attach failed (Not allowed to attach to process.  Look in the console messages (Console.app), near the debugserver entries, when the attach failed.  The subsystem that denied the attach permission will likely have logged an informative message about why it was denied.))
```

console.app says: `macOSTaskPolicy: (com.apple.debugserver) may not get the task control port of (bun-index) (pid: 40225): (bun-index) is hardened, (bun-index) doesn't have get-task-allow, (com.apple.debugserver) is a declared debugger(com.apple.debugserver) is not a declared read-only debugger`, whatever that means

**Comment by llimllib** [comment]
Updated for 1.0: it doesn't segfault, but it does exit with `FileNotFound` and doesn't tell me what file it failed to find:

```
$ bun build --compile src/index.ts --outfile bun-index --target=node
  [45ms]  bundle  993 modules
  [96ms] compile  bun-index
$ ./bun-index --help                                                


error: FileNotFound reading "node:path"
```

**Reply by llimllib** [reply]
`process.env.NODE_CONFIG_DIR = path.join(__dirname, '../config');` seems to be the line that's killing it

**Reply by llimllib** [reply]
filed as https://github.com/oven-sh/bun/issues/4592
